### PR TITLE
Predict upgrade

### PR DESCRIPTION
**Pull request by bonham79**
For https://github.com/CUNY-CL/yoyodyne/issues/26

Major changes:

-- base encoder now contains predict_step that applies to lstm and transformer models.
--Put wrapper around base_encoder.__get_loss method (explained below)

@Adamits  @kylebgorman  This is a minor fix but want to get y'alls thoughts on best practices: While going through this, it looks like we were using two logics for forward passes. For training, tensors were assumed to be of shape `B x vocab x sequence_len` while on prediction they were assumed as `B x sequence_len x vocab` . This isn't ideal and led to a bunch of transpose steps that were somewhat unnecessary. I've tried to unify them under the assumption that alll `forward` passes predict a tensor of shape `B x sequence_len x vocab`

Channeling my innner Guido I want to suggest the following:
1. All `forward` passes (unless in specialized cases), produce tensors of one unified shape. My preference is for `B x seq_len x vocab_size` but am open to other arguments.
2. `base_encoder._get_predictions` should only be called on decoder outputs. Since this is really only a wrapper for `max`, I think it's more important to save semantic sanity than using it for condensing code. 

Anyway, these are just my thoughts.

### CODE REVIEW

**Comment by kylebgorman** [comment]
> @Adamits @kylebgorman This is a minor fix but want to get y'alls thoughts on best practices: While going through this, it looks like we were using two logics for forward passes. For training, tensors were assumed to be of shape `B x vocab x sequence_len` while on prediction they were assumed as `B x sequence_len x vocab` . This isn't ideal and led to a bunch of transpose steps that were somewhat unnecessary. I've tried to unify them under the assumption that alll `forward` passes predict a tensor of shape `B x sequence_len x vocab`

Oof, is that so? Glad you fixed. 

> Channeling my innner Guido I want to suggest the following:
> 
> 1. All `forward` passes (unless in specialized cases), produce tensors of one unified shape. My preference is for `B x seq_len x vocab_size` but am open to other arguments.

No preference so long as it's consistent and batch size is the leading dimension.

File: yoyodyne/indexes.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 """Indexes for symbols."""
 
-import pickle
 import os
+import pickle
```
Comment (kylebgorman) [inline yoyodyne/indexes.py:4]: please revert

File: yoyodyne/models/base.py [hunk 2]
```diff
@@ -235,6 +235,28 @@ def validation_epoch_end(self, validation_step_outputs: Dict) -> Dict:
...
+        Args:
+            batch (batches.PaddedBatch).
+            batch_idx (int).
+
+        Returns:
+            torch.Tensor: indices of the argmax at each timestep.
+        """
+        self.eval()
+        predictions = self(batch)
+        # Default tensor: B x seq_len x vocab_size.
```
Comment (kylebgorman) [inline yoyodyne/models/base.py:256]: What does "Default tensor" mean? Is this just saying the next step returns something of shape B x seq_len x vocab_size? If so then use Adam's notation:

`# -> B x seq_len x vocab_size.`

File: yoyodyne/models/transducer.py [hunk 3]
```diff
@@ -45,20 +45,21 @@ def __init__(
...
-        # Alternate outputs than dataset targets.
-        kwargs["output_size"] = len(expert.actions)
-        super().__init__(*args, **kwargs)
         # Target embeddings use alternate padding.
         self.target_embeddings = self.init_embeddings(
             num_embeddings=self.output_size,
             embedding_size=self.embedding_size,
             pad_idx=self.actions.end_idx,
         )
+        self.loss_func = TransducerNoFeatures.loss_func
```
Comment (kylebgorman) [inline yoyodyne/models/transducer.py:62]: This requires documentation. You're trying to say that you're going to overload it later? Then this should be no-op. Or are you trying to say something else?

**Review by kylebgorman — COMMENTED** [review]
This is nice for getting rid of the stuff in predict! 

A few nits/comments (see also main thread). 

Can you standardize on either `seq_len` or (and I prefer this, particularly when these are just comments anyways) `sequence_length` for

**Comment by kylebgorman** [comment]
> > 1. All `forward` passes (unless in specialized cases), produce tensors of one unified shape. My preference is for `B x seq_len x vocab_size` but am open to other arguments.
> 
> No preference so long as it's consistent and batch size is the leading dimension.

Revising my comment after reviewing. If the loss function wants it one way, we should just do it that way everywhere unless there's some other built-in that wants it transposed.

File: yoyodyne/models/transducer.py (hunk 3 above)
Comment (bonham79) [reply yoyodyne/models/transducer.py:62]: I'm having a hard time making this non-obtrusive. The transducer forward pass computes a loss with it, so its loss function is just a dummy wrapper for referencing a tuple. Current inheritance order wants to call `base`'s loss function finder so I'm trying to override that function.

**Comment by bonham79** [comment]
> Revising my comment after reviewing. If the loss function wants it one way, we should just do it that way everywhere unless there's some other built-in that wants it transposed.

That can be done, but many of the forward passes have a default of `B x seq_len x vocab_size`, so doing the other way adds in quite a few `transpose` operations. It also means `base._get_predictions` is more limited in when it can be called (to make semantic sense, it would have to operate on a different dimension, so can't be called with some of the `lstm` outputs until bottom of the forward pass. 

`transpose` is a cheap operation iirc, but I personally find too many of them hard to follow on a code edit. If this is just a me problem I'm fine with the other way.

File: yoyodyne/models/transducer.py (hunk 3 above)
Comment (kylebgorman) [reply yoyodyne/models/transducer.py:62]: Okay, well, you need to write a clear, sentential comment when you use a strange pattern: explain what the pattern does functionally, and why you want to do it.

Secondly, my comments otherwise suggest: we should structure batches to follow what the loss function wants and do away with transposition altogether.

**Comment by kylebgorman** [comment]
> That can be done, but many of the forward passes have a default of `B x seq_len x vocab_size`, 

Can't we just change this too? If so we should. 

> `transpose` is a cheap operation iirc, but I personally find too many of them hard to follow on a code edit. If this is just a me problem I'm fine with the other way.

No you're right, they're awful.

**Comment by bonham79** [comment]
> 
> Can't we just change this too? If so we should.
> 
Nah, this is a product of the pytorch operations. Particularly how lstms have a weird output behavior. Adam did a good job and still you have weirdness like `tensor.tranpose(0,1).tranpose(1,2)`

File: yoyodyne/indexes.py (hunk 1 above)
Comment (bonham79) [reply yoyodyne/indexes.py:4]: line 65 `dictionary = pickle.load(source)` line 77 `pickle.dump(vars(self), sink)` ?

Comment (kylebgorman) [reply yoyodyne/indexes.py:4]: No, lines 2, and 4. You changed the order of imports (pickle was already being imported) to be non-lexicographic.

Comment (bonham79) [reply yoyodyne/indexes.py:4]: oh gotcha, isort was overzealous

**Comment by bonham79** [comment]
Ignore changes to unrelated files, my linters are having issues for some reason

**Review by kylebgorman — APPROVED** [review]
LGTM. @Adamits do you want to take a look?

File: yoyodyne/models/base.py [hunk 4]
```diff
@@ -235,11 +235,33 @@ def validation_epoch_end(self, validation_step_outputs: Dict) -> Dict:
...
+        Args:
+            batch (batches.PaddedBatch).
+            batch_idx (int).
+
+        Returns:
+            torch.Tensor: indices of the argmax at each timestep.
+        """
+        self.eval()
+        predictions = self(batch)
+        # -> B x seq_len x output_size.
```
Comment (Adamits) [inline yoyodyne/models/base.py:256]: Are you sure this doesn't return `B x seq_len x 1`?

**Comment by Adamits** [comment]
Started reviewing and realized that I am pretty sure there is overlap with #46, which I reviewed first. Should I look at all commits here?

File: yoyodyne/models/lstm.py [hunk 5]
```diff
@@ -339,8 +339,8 @@ def beam_decode(
                 char_likelihoods,
                 decoder_hiddens,
             ) in likelihoods:
-                # -> B x seq_len x outputs.
-                # This is 1 x 1 x outputs since we fixed batch size to 1.
+                # -> B x seq_len x output_size.
```
Comment (Adamits) [inline yoyodyne/models/lstm.py:342]: Maybe just remove this comment.
The one below (sans "this is") should be sufficient.

**Comment by kylebgorman** [comment]
#46 is just a superset of this, I believe; it has one more commit than this one (#44).

File: yoyodyne/models/transducer.py [hunk 6]
```diff
@@ -488,19 +488,16 @@ def log_sum_softmax_loss(
...
 
         This model already produces prediction strings and calculates loss
-        during training. This simply serves as a wrapper to allow compatibility
-        with training_step in the superclass.
+        during training. This simply serves as a wrapper to prevent
+        superclass from creating an unused loss function.
         """
-        _, loss = prediction
-        return loss
+        return lambda predictions, target: predictions[1]
```
Comment (Adamits) [inline yoyodyne/models/transducer.py:500]: Not sure how this is usually handled, but it strikes me as a bit odd that this is conceptually returns a loss function, but we have no docstring for it. If I am a developer making changes to `Transducer` is it clear how to interact with the loss function?

File: yoyodyne/models/transformer.py [hunk 7]
```diff
@@ -212,17 +212,17 @@ def _decode_greedy(
             # We only care about the last prediction in the sequence.
             last_output = output[:, -1, :]
             outputs.append(last_output)
-            pred = self._get_predicted(last_output.unsqueeze(1))
-            predictions.append(pred.squeeze(1))
+            _, pred = torch.max(last_output, dim=1)
```
Comment (Adamits) [inline yoyodyne/models/transformer.py:215]: A comment could be good here:
```
# -> B x 1 x 1
```

File: yoyodyne/models/base.py [hunk 8]
```diff
@@ -323,7 +345,31 @@ def get_loss_func(
                 loss function.
         """
         if self.label_smoothing is None:
-            return nn.NLLLoss(ignore_index=self.pad_idx, reduction=reduction)
+            loss_func = nn.NLLLoss(
+                ignore_index=self.pad_idx, reduction=reduction
+            )
+
+            def _transpose_nlloss(
```
Comment (Adamits) [inline yoyodyne/models/base.py:352]: This is all a bit weird, but pretty sure it's updated in the next PR!

File: yoyodyne/models/base.py (hunk 4 above)
Comment (bonham79) [reply yoyodyne/models/base.py:256]: Oh, mb.

File: yoyodyne/models/transducer.py (hunk 6 above)
Comment (bonham79) [reply yoyodyne/models/transducer.py:500]: This should be more coherrent in the other PR. Lemme know if you still see issues. Dev work should never touch the loss function since it's part o the forward pass.

**Comment by bonham79** [comment]
Adams recs are done in the other PR to avoid commit conflicts.

**Comment by kylebgorman** [comment]
> Adams recs are done in the other PR to avoid commit conflicts.

So does that mean this is ready?
